lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php

```diff
@@ -0,0 +1,611 @@
+<?php
+/**
+ * WP_Navigation_Block_Renderer class
+ *
+ * @package gutenberg
+ * @since 6.5.0
+ */
+
+if ( class_exists( 'WP_Navigation_Block_Renderer' ) ) {
+	return;
+}
+
+/**
+ * Helper functions used to render the navigation block.
+ */
+class WP_Navigation_Block_Renderer {
+
+	private static $list_item_nav_blocks = array(
+		'core/navigation-link',
+		'core/home-link',
+		'core/site-title',
+		'core/site-logo',
+		'core/navigation-submenu',
+	);
+
+	private static $needs_list_item_wrapper = array(
+		'core/site-title',
+		'core/site-logo',
+	);
+
+	/**
+	 * Returns whether or not this is responsive navigation.
+	 *
+	 * @param array $attributes The block attributes.
+	 * @return bool Returns whether or not this is responsive navigation.
+	 */
+	private static function is_responsive( $attributes ) {
+		/**
+		 * This is for backwards compatibility after the `isResponsive` attribute was been removed.
+		 */
+
+		$has_old_responsive_attribute = ! empty( $attributes['isResponsive'] ) && $attributes['isResponsive'];
+		return isset( $attributes['overlayMenu'] ) && 'never' !== $attributes['overlayMenu'] || $has_old_responsive_attribute;
+	}
+
+	/**
+	 * Returns whether or not a navigation has a submenu.
+	 *
+	 * @param WP_Block_List $inner_blocks The list of inner blocks.
+	 * @return bool Returns whether or not a navigation has a submenu.
+	 */
+	private static function has_submenus( $inner_blocks ) {
+		$has_submenus = false;
+		foreach ( $inner_blocks as $inner_block ) {
+			$inner_block_content = $inner_block->render();
+			$p                   = new WP_HTML_Tag_Processor( $inner_block_content );
+			if ( $p->next_tag(
+				array(
+					'name'       => 'LI',
+					'class_name' => 'has-child',
+				)
+			) ) {
+				$has_submenus = true;
+			}
```

Comment:
```suggestion
				$has_submenus = true;
				break;
			}
```

Could we exit the loop early once we've determine there are submenus?

Reply:
done in 939c50ff52